### PR TITLE
Fixes for IGT crashes

### PR DESCRIPTION
**Pull request by mripard**
Hi,

Here's a collection of fixes related to the IGT crashes @popcornmix reported. It turns out many of them were actually due to the fact that even though the BCM2711 doesn't have the v3d handled by the vc4 driver anymore, we were still using some of the code backed by the v3d.

This PR properly splits everything away, and adds a warning to all the v3d-related entrypoints so that we notice if that it ever happens again.

### CODE REVIEW

File: drivers/gpu/drm/vc4/vc4_kms.c [hunk 1]
```diff
@@ -395,7 +395,7 @@ static void vc4_atomic_commit_tail(struct drm_atomic_state *state)
 		old_hvs_state->fifo_state[channel].pending_commit = NULL;
 	}
 
-	if (vc4->hvs && vc4->hvs->hvs5) {
+	if (vc4->is_vc5) {
```
Comment (popcornmix) [inline drivers/gpu/drm/vc4/vc4_kms.c]: This breaks fkms (where `vc4->hvs` would be null). Suggest `if (vc4->is_vc5 && !vc4->firmware_kms)` as fix.

File: drivers/gpu/drm/vc4/vc4_kms.c [hunk 2]
```diff
@@ -427,7 +427,7 @@ static void vc4_atomic_commit_tail(struct drm_atomic_state *state)
 
 	drm_atomic_helper_cleanup_planes(dev, state);
 
-	if (vc4->hvs && vc4->hvs->hvs5) {
+	if (vc4->is_vc5) {
```
Comment (popcornmix) [inline drivers/gpu/drm/vc4/vc4_kms.c]: This breaks fkms (where `vc4->hvs` would be null). Suggest `if (vc4->is_vc5 && !vc4->firmware_kms)` as fix.

**Comment by popcornmix** [comment]
I've been running with this on Pi4 with kms over weekend without issue.
fkms threw up a couple a null pointer dereferences (annoying this always occurs in the dead time after uart output stops but before you can log in to see dmesg), but I've suggested fixes for them.
Booted with kms and fkms on PI2.
No issues spotted reading commits.

**Comment by 6by9** [comment]
>  (annoying this always occurs in the dead time after uart output stops but before you can log in to see dmesg),

Blacklist ```vc4```. Reboot. Wait for the system to come up. SSH in and run ```sudo modprobe vc4```. You may need a second SSH session to be able to view the logs.

**Comment by mripard** [comment]
Thanks for your comments, I'll address them.

> > (annoying this always occurs in the dead time after uart output stops but before you can log in to see dmesg),
> 
> Blacklist `vc4`. Reboot. Wait for the system to come up. SSH in and run `sudo modprobe vc4`. You may need a second SSH session to be able to view the logs.

Alternatively, we can disable the fbdev emulation in Kconfig (CONFIG_DRM_FBDEV_EMULATION), and then try to set a mode by hand.

**Comment by popcornmix** [comment]
> Blacklist vc4. Reboot. Wait for the system to come up. SSH in and run sudo modprobe vc4

Excellent. That works for finding this hang.

> Alternatively, we can disable the fbdev emulation in Kconfig (CONFIG_DRM_FBDEV_EMULATION), and then try to set a mode by hand.

Yes, this also works.

My scheme was adding a `v7_flush_kern_dcache_area` after `prb_final_commit` then using VideoCore debugger to dump memory from address of `__log_buf` (found from `System.map`). But both of these are simpler.

**Comment by 6by9** [comment]
Spot those of us who have also been hunting this style of NULL deref or lockup in the past :-)

**Comment by pelwell** [comment]
I think this is just waiting for the fkms-proofing changes that @popcornmix suggested.

**Comment by mripard** [comment]
I just pushed a new version with the changes @popcornmix requested

**Comment by pelwell** [comment]
Thanks!

**Pull request merged**